jax/experimental/sparse/linalg.py

```diff
@@ -0,0 +1,388 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Sparse linear algebra routines."""
+
+from typing import Union, Callable
+import functools
+
+import jax
+import jax.numpy as jnp
+
+import numpy as np
+
+def lobpcg_standard(
+    A: Union[jnp.ndarray, Callable[[jnp.ndarray], jnp.ndarray]],
+    X: jnp.ndarray,
+    m: int = 100,
+    tol: Union[jnp.ndarray, float, None] = None):
+  """Compute the top-k standard eigenvalues using the LOBPCG routine.
+
+  LOBPCG [1] stands for Locally Optimal Block Preconditioned Conjugate Gradient.
+  The method enables finding top-k eigenvectors in an accelerator-friendly
+  manner.
+
+  This initial experimental version has several caveats.
+
+    - Only the standard eigenvalue problem `A U = lambda U` is supported,
```

Comment:
This MR seems a little odd to me. The power of LOBPCG is primarily on finding the smallest eigenvalues when a preconditioner for A is known. For finding the largest, the (block) Lanczos algorithm is provably optimal and well-understood.

Reply:
Discussed offline: this PR adds just top-k to make the first change as small as we can get.

The extra preconditioner param and support for k-smallest eigenvals will come in a follow up PR(s) which will help reduce reviewer load for testing/locs etc.